protocols/kad/src/behaviour/test.rs

```diff
@@ -695,6 +695,407 @@ fn put_record() {
     QuickCheck::new().tests(3).quickcheck(prop as fn(_, _) -> _)
 }
 
+/// A node joining a fully connected network cannot store a record if the nodes ignore it through
```

Comment:
A bit late to suggest this, sorry about that. Still: How about integrating this test into `put_record` by adding something along the lines of:

```
        if rng.gen() {
             config.set_record_filtering(KademliaStoreInserts::FilterBoth);
        }
```

And then, in case one receives a `InboundPutRecordRequest`, double check that `config.record_filtering` was set to `FilterBoth`.

That would reduce the amount of code required, no?

Reply:
Do you have some coverage analysis in place to check that these code paths are hit? Using an rng for that sounds scary :'-)

I'll implement it like that though, sounds a lot cleaner. I'll test it using provider records too.

Reply:
You are right. Let's get rid of the RNG. How about the diff below + optionally setting `set_record_filtering` in the `TestConfig` `Arbitrary` implementation?

```diff
diff --git a/protocols/kad/src/behaviour/test.rs b/protocols/kad/src/behaviour/test.rs
index 110c1e22..b27fd076 100644
--- a/protocols/kad/src/behaviour/test.rs
+++ b/protocols/kad/src/behaviour/test.rs
@@ -488,24 +488,28 @@ fn get_record_not_found() {
 /// is equal to the configured replication factor.
 #[test]
 fn put_record() {
-    fn prop(records: Vec<Record>, seed: Seed) {
-        let mut rng = StdRng::from_seed(seed.0);
-        let replication_factor =
-            NonZeroUsize::new(rng.gen_range(1, (K_VALUE.get() / 2) + 1)).unwrap();
-        // At least 4 nodes, 1 under test + 3 bootnodes.
-        let num_total = usize::max(4, replication_factor.get() * 2);
-
-        let mut config = KademliaConfig::default();
-        config.set_replication_factor(replication_factor);
-        if rng.gen() {
-            config.disjoint_query_paths(true);
+    #[derive(Clone, Debug)]
+    struct TestConfig(KademliaConfig);
+
+    impl Arbitrary for TestConfig {
+        fn arbitrary<G: Gen>(g: &mut G) -> TestConfig {
+            let mut config = KademliaConfig::default();
+            config.set_replication_factor(NonZeroUsize::new(g.gen_range(1, (K_VALUE.get() / 2) + 1)).unwrap());
+            config.disjoint_query_paths(g.gen());
+            TestConfig(config)
         }
+    }
+
+    fn prop(records: Vec<Record>, config: TestConfig) {
+        let config = config.0;
+        // At least 4 nodes, 1 under test + 3 bootnodes.
+        let num_total = usize::max(4, config.query_config.replication_factor.get() * 2);
 
         let mut swarms = {
             let mut fully_connected_swarms =
                 build_fully_connected_nodes_with_config(num_total - 1, config.clone());
 
-            let mut single_swarm = build_node_with_config(config);
+            let mut single_swarm = build_node_with_config(config.clone());
             // Connect `single_swarm` to three bootnodes.
             for i in 0..3 {
                 single_swarm.1.behaviour_mut().add_address(
@@ -584,7 +588,7 @@ fn put_record() {
                         ))) => {
                             assert!(qids.is_empty() || qids.remove(&id));
                             assert!(stats.duration().is_some());
-                            assert!(stats.num_successes() >= replication_factor.get() as u32);
+                            assert!(stats.num_successes() >= config.query_config.replication_factor.get() as u32);
                             assert!(stats.num_requests() >= stats.num_successes());
                             assert_eq!(stats.num_failures(), 0);
                             match res {
@@ -635,7 +639,7 @@ fn put_record() {
 
                 let expected = expected
                     .into_iter()
-                    .take(replication_factor.get())
+                    .take(config.query_config.replication_factor.get())
                     .collect::<HashSet<_>>();
 
                 let actual = swarms
@@ -650,7 +654,7 @@ fn put_record() {
                     })
                     .collect::<HashSet<_>>();
 
-                assert_eq!(actual.len(), replication_factor.get());
+                assert_eq!(actual.len(), config.query_config.replication_factor.get());
 
                 let actual_not_expected = actual.difference(&expected).collect::<Vec<&PeerId>>();
                 assert!(
(END)

```

Reply:
With only 3 tests, I have a test case that doesn't get hit. With four, it does seem to hit all paths. I'll add two boolean parameters to the test properties instead, although I still have to increase the test count.